src/NuGet.Core/NuGet.Packaging/Signing/Archive/SignedPackageArchiveIOUtility.cs

```diff
@@ -139,95 +175,131 @@ public static void HashBytes(HashAlgorithm hashAlgorithm, byte[] bytes)
+                    hasFoundSignature = true;
+                }
 
-                    // The total size of file entry is from the start of the file header until
-                    // the start of the next file header (or the start of the first central directory header)
-                    try
-                    {
-                        SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(LocalFileHeaderSignature));
-                    }
-                    // No local File header found (the signature file must be the last entry), search for the start of the first central directory
-                    catch
-                    {
-                        SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(CentralDirectoryHeaderSignature));
-                    }
+                // Go to local file header
+                reader.BaseStream.Seek(offset: record.OffsetToFileHeader, origin: SeekOrigin.Begin);
 
-                    metadata.SignatureFileEntryTotalSize = reader.BaseStream.Position - metadata.SignatureLocalFileHeaderPosition;
+                // Validate file header signature
+                var fileHeaderSignature = reader.ReadUInt32();
+                if (fileHeaderSignature != LocalFileHeaderSignature)
+                {
+                    throw new InvalidDataException(Strings.ErrorInvalidPackageArchive);
                 }
 
+                // The total size of file entry is from the start of the file header until
+                // the start of the next file header (or the start of the first central directory header)
                 try
                 {
-                    SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(CentralDirectoryHeaderSignature));
-                    possibleSignatureCentralDirectoryRecordPosition = reader.BaseStream.Position;
+                    SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(LocalFileHeaderSignature));
                 }
+                // No local File header found (entry must be the last entry), search for the start of the first central directory
                 catch
                 {
-                    break;
+                    SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(CentralDirectoryHeaderSignature));
                 }
+
+                record.IndexInRecords = centralDirectoryRecordIndex;
+                record.FileEntryTotalSize = reader.BaseStream.Position - record.OffsetToFileHeader;
             }
 
-            if (!hasFoundSignatureEntry)
+            if (!hasFoundSignature)
             {
-                throw new SignatureException(Strings.ErrorPackageNotSigned);
+                throw new SignatureException(Strings.Error_NotOnePrimarySignature);
             }
 
+            metadata.CentralDirectoryRecords = centralDirectoryRecords;
+
+            // Make sure the package is not zip64
```

Comment:
If I throw an exception inside the try/catch is going to be catched and then ignored. I used the Boolean because `SeekReaderForwardToMatchByteSignature` will throw if the signature is not found, so if that doesn't throw then I want to throw, but if it throws I just want to ignore it and continue. (if it throws it means is not zip64)

Reply:
my bad!